---
title: "efo_enrichments"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(ontologyIndex)
library(tidyverse)
library(rtracklayer)
library(plyranges)

```

First pass calculating LD for all chromosomes.

# Read in all LD files

```{r}
ld_files_list <- list.files("../data_output/plink/gwas_ld_snps_27jan22/",pattern = "ld_snps.ld",full.names = TRUE)

out_ld_df <- do.call(rbind, lapply(ld_files_list, read_table))

gc()
```

# Filter by R2 > 0.7

Filtering for LD with R2 of at least 0.7
```{r}
out_ld_df <- out_ld_df %>%
  dplyr::filter(R2 >= 0.7)

gc()
```

# Make GRanges for overlaps

```{r}
out_ld_gr <- makeGRangesFromDataFrame(df = out_ld_df,
                         seqnames.field = "CHR_B",
                         start.field = "BP_B",
                         end.field = "BP_B",
                         keep.extra.columns = TRUE)
```

Read haqers
```{r }
haqers <- import.bed("../data_raw/bed/HAQER.bed")

haqers$name <- paste0("haqer",1:length(haqers))
seqlevels(haqers) <- str_remove(seqlevels(haqers) , "chr")


```



Read random regions
```{r}
read_haqers_like_bed <- function(file){
  haqers <- import.bed(file)
  haqers$name <- paste0("haqer_rand",1:length(haqers))
  seqlevels(haqers) <- str_remove(seqlevels(haqers) , "chr")
  return(haqers)
}

rand_haqers_list <- lapply(list.files("../data_output/bed/random_beds",full.names = TRUE)[1:100], read_haqers_like_bed)
```


# Number of overlaps

Overlap haqers and ld gwas snps
```{r}
gwas_ld_snp_haqers_gr <- plyranges::join_overlap_inner(haqers,out_ld_gr)

```

Number of haqers that overlap at least one GWAS ld snp
```{r}
n_olaps_haqers <- length(unique(gwas_ld_snp_haqers_gr$name))
```

Overlap random haqers and ld gwas snps
```{r}
gwas_ld_snp_random_gr_list <- lapply(rand_haqers_list, plyranges::join_overlap_inner, y = out_ld_gr)
```

Number of random haqers that overlap at least one GWAS ld snp
```{r}
n_rand_olaps <- unlist(lapply(gwas_ld_snp_random_gr_list, function(x) {length(unique(x$name))}))
```

Histogram of number of random haqers that overlap at least one gwas ld snp 
```{r}
n_rand_olaps %>%
  enframe() %>%
  ggplot(aes(value)) +
  geom_histogram(color = "white") +
  theme_light() +
  theme(aspect.ratio = 1) +
  geom_vline(xintercept = n_olaps_haqers,color = "red", lty = "dashed")
```


# Enrichment of EFO classes


Read gwas catalog information
```{r}
gwas_catalog_snp <- read_tsv("../data_raw/gwas_catalog/gwas_catalog_v1.0.2-associations_e105_r2021-12-21.tsv")
# Change types of some variables to make them compatible with later table join
gwas_catalog_snp$CHR_POS <- as.numeric(gwas_catalog_snp$CHR_POS)
gwas_catalog_snp <- gwas_catalog_snp[!is.na(gwas_catalog_snp$CHR_POS),]

```

Change mapped trait variable to EFO class

Map broad Experimental Factor Ontology categories to phenotypes to do enrichment of broad disease phenotypes (e.g. brain disease etc.)

```{r}
efo <- get_ontology("../data_raw/obo/efo.obo")

```


## Get ancestor ontology terms for each mapped trait in gwas
```{r}
gwas_catalog_snp$mapped_trait_fct <- as.factor(gwas_catalog_snp$MAPPED_TRAIT)


# Get mapped trait efo IDs

mapped_trait_efo_ids <- efo$name[efo$name %in% levels(gwas_catalog_snp$mapped_trait_fct)]

# Get ancestors for each mapped trait
mapped_trait_efo_ancestors <- map(names(mapped_trait_efo_ids), get_term_property,ontology=efo, property="ancestors", as_names=TRUE)
names(mapped_trait_efo_ancestors) <- mapped_trait_efo_ids



```

## Filter, get only those with disease traits
```{r}
# Get mapped traits with "disease" as an ancestor
disease_efo_id <- efo$name[efo$name %in% "disease"]

mapped_trait_efo_ancestors_disease <- map_lgl(mapped_trait_efo_ancestors, function(x) { 
  return(any(x %in% disease_efo_id))
})



gwas_catalog_snp <- gwas_catalog_snp[gwas_catalog_snp$MAPPED_TRAIT %in% mapped_trait_efo_ids[mapped_trait_efo_ancestors_disease],]
```

## Get child term for each disease trait
```{r}
gwas_catalog_snp$mapped_trait_fct <- fct_drop(gwas_catalog_snp$mapped_trait_fct)

mapped_trait_efo_ancestors <- mapped_trait_efo_ancestors[levels(gwas_catalog_snp$mapped_trait_fct)]

mapped_trait_efo_disease <- map_chr(mapped_trait_efo_ancestors, function(x) {
  disease_id_pos <- which(x == "disease")
  return(x[disease_id_pos+1])
} )

mapped_trait_efo_disease_old <- names(mapped_trait_efo_disease)
mapped_trait_efo_disease_new <- mapped_trait_efo_disease
names(mapped_trait_efo_disease_old) <- mapped_trait_efo_disease_new

gwas_catalog_snp$MAPPED_TRAIT <- fct_recode(gwas_catalog_snp$mapped_trait_fct, !!!mapped_trait_efo_disease_old)
```



Overlap gwas phenotype info with haqers that overlap a gwas snp, join by coordinate of gwas snp

```{r}
# Join a GRanges of haqers, with metadata containing the gwas and ld snps they overlap, and a gwas phenotype info table, join by coordinate of gwas snp
get_overlaps_phenos <- function(haqers_overlaps, gwas_snp_info) {
  haqers_overlaps <- as.data.frame(haqers_overlaps)
  haqers_overlaps$CHR_A <- as.character(haqers_overlaps$CHR_A)
  haqers_overlaps$BP_A <- as.numeric(haqers_overlaps$BP_A)
  
  haqers_overlaps_pheno <- left_join(haqers_overlaps,gwas_snp_info, by = c("CHR_A"="CHR_ID", "BP_A"="CHR_POS"))
  return(haqers_overlaps_pheno)
}

# Get the Mapped trait column from the joined haqers - gwas snp - phenotype table, and tabulate the number of haqers associated with each phenotype
get_haqers_per_pheno <- function(.x,.y) {
  haqers_per_pheno <- as.data.frame(table(unique(.x[,c("name","MAPPED_TRAIT")])$MAPPED_TRAIT))
  haqers_per_pheno$n_rand <- .y
  return(haqers_per_pheno)
}
```

Join random haqers-snps granges with gwas phenotype info 
```{r}
gwas_ld_snp_random_pheno_df_list <- lapply(gwas_ld_snp_random_gr_list,get_overlaps_phenos,gwas_snp_info = gwas_catalog_snp)

```

Get number of haqers per phenotype for each random haqer
```{r}

names(gwas_ld_snp_random_pheno_df_list) <- paste0("rand_",1:length(gwas_ld_snp_random_pheno_df_list))
haqers_per_pheno_rand_list <- imap(gwas_ld_snp_random_pheno_df_list,get_haqers_per_pheno)
```

Bind all 
```{r}
haqers_per_pheno_rand_df <- do.call(rbind,haqers_per_pheno_rand_list)
colnames(haqers_per_pheno_rand_df) <- c("mapped_trait","counts","id_rand")

```

Get same info for actual haqers

```{r}
haqers_olap_pheno_df <- get_overlaps_phenos(gwas_ld_snp_haqers_gr,gwas_catalog_snp)
haqers_per_pheno_counts <- get_haqers_per_pheno(haqers_olap_pheno_df,"haqers")

colnames(haqers_per_pheno_counts) <- c("mapped_trait","counts","id_haqer")
haqers_per_pheno_counts <- select(haqers_per_pheno_counts, -id_haqer)

```

Get "p value" per phenotype

```{r}
# Convert factors to characters for comparison
haqers_per_pheno_rand_df$mapped_trait <- as.character(haqers_per_pheno_rand_df$mapped_trait)
haqers_per_pheno_counts$mapped_trait <- as.character(haqers_per_pheno_counts$mapped_trait)


get_pheno_pval <- function(.x, .y, random_counts_df){
  random_counts <- random_counts_df[random_counts_df$mapped_trait == .x,"counts"]
  nrand <- length(unique(random_counts_df$id_rand))
  pval <- sum(random_counts >= .y ) / nrand
  return(pval)
}

haqers_per_pheno_counts$pval <- map2_dbl(as.character(haqers_per_pheno_counts$mapped_trait),
         haqers_per_pheno_counts$counts,
         get_pheno_pval,
         random_counts_df = haqers_per_pheno_rand_df)

haqers_per_pheno_counts$fdr <- p.adjust(haqers_per_pheno_counts$pval,method = "fdr")
```

Distribution of p values
```{r}
haqers_per_pheno_counts %>%
  ggplot(aes(pval)) +
  geom_histogram()
```
Phenotypes
```{r}
haqers_per_pheno_counts %>%
  filter(counts > 1) %>%
  mutate(fdr = p.adjust(pval)) %>%
  arrange(fdr)
```
```{r}
trait <- "respiratory system disease"

haqers_per_pheno_rand_df %>%
  filter(mapped_trait == trait) %>%
  ggplot(aes(counts)) +
  geom_histogram() +
  geom_vline(xintercept = haqers_per_pheno_counts %>%
               filter(mapped_trait == trait) %>%
               pull(counts),
             lty = "dashed",
             color = "red")+
  ggtitle(trait) +
  theme_light()
```


```{r}
trait <- "digestive system disease"

haqers_per_pheno_rand_df %>%
  filter(mapped_trait == trait) %>%
  ggplot(aes(counts)) +
  geom_histogram() +
  geom_vline(xintercept = haqers_per_pheno_counts %>%
               filter(mapped_trait == trait) %>%
               pull(counts),
             lty = "dashed",
             color = "red") +
  ggtitle(trait) +
  theme_light()

```




