---
title: "First pass of intersecting haqers with gwas + ld variants using chr22 only"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries
```{r}
library(tidyverse)
library(rtracklayer)
library(plyranges)
options(scipen = 9999)
```

# Read and preprocess inputs

## HAQERs

Read haqers
```{r }
haqers <- import.bed("../data_raw/bed/HAQER.bed")

haqers$name <- paste0("haqer",1:length(haqers))
seqlevels(haqers) <- str_remove(seqlevels(haqers) , "chr")

haqers <- haqers[seqnames(haqers) == 22]
```

## GWAS catalog

```{r}
gwas_catalog_snp <- read_tsv("../data_raw/gwas_catalog/gwas_catalog_v1.0.2-associations_e105_r2021-12-21.tsv")
gwas_catalog_populations <- read_tsv("../data_raw/gwas_catalog/gwas_catalog-ancestry_r2021-12-21.tsv")
```

Note: some associations are filtered out here due to them not having an associated snp.
```{r}

gwas_catalog_snp <- gwas_catalog_snp[!is.na(gwas_catalog_snp$CHR_POS ),]

gwas_catalog_snp$CHR_POS_NUMERIC <- as.numeric(gwas_catalog_snp$CHR_POS)


gwas_catalog_snp <- gwas_catalog_snp[!is.na(gwas_catalog_snp$CHR_POS_NUMERIC),]
```



Filter by population, get SNPs ascertained in European populations 
```{r}
gwas_catalog_populations_european <- gwas_catalog_populations[gwas_catalog_populations$`BROAD ANCESTRAL CATEGORY`=="European",c("STUDY ACCESSION"),]
gwas_catalog_populations_european <- unique(gwas_catalog_populations_european)

```

```{r}
gwas_catalog_snp <- gwas_catalog_snp[gwas_catalog_snp$`STUDY ACCESSION` %in% gwas_catalog_populations_european$`STUDY ACCESSION`,]
```

Filter by chromosome, get chr22 

```{r}
gwas_catalog_snp <- gwas_catalog_snp[gwas_catalog_snp$CHR_ID == "22",]
```

Filter out multi snp records from gwas catalog
```{r}
gwas_catalog_snp <- gwas_catalog_snp[!is.na(gwas_catalog_snp$SNP_ID_CURRENT),]
```

Make granges for overlaps
```{r}
gwas_catalog_snp_gr_chr22 <- makeGRangesFromDataFrame(gwas_catalog_snp,
                         seqnames.field = "CHR_ID",
                         start.field = "CHR_POS_NUMERIC",
                         end.field = "CHR_POS_NUMERIC",
                         keep.extra.columns = TRUE
                         )
```


How to handle the fact that there could be different traits associated to the same snp.
Get a list of unique ranges:
Keep SNPs column for later join with phenotype info
```{r}
gwas_catalog_snp_gr_chr22_uq <- gwas_catalog_snp_gr_chr22[,"SNPS"]
gwas_catalog_snp_gr_chr22_uq <- unique(gwas_catalog_snp_gr_chr22_uq)
```



## 1000 genomes

Cut vcf to only read first three columns
```{bash eval = F}
gzip -cd ../data_raw/genomes_1000/CCDG_14151_B01_GRM_WGS_2020-08-05_chr22.filtered.shapeit2-duohmm-phased.vcf.gz |grep -v '^##'| cut -f1-3 > ../data_output/genomes_1000/CCDG_14151_B01_GRM_WGS_2020-08-05_chr22.filtered.shapeit2-duohmm-phased.snp_id.txt
```

```{r }
chr22_phased_highcov_phase3 <- read_table("../data_output/genomes_1000/CCDG_14151_B01_GRM_WGS_2020-08-05_chr22.filtered.shapeit2-duohmm-phased.snp_id.txt")
```

```{r}
chr22_phased_highcov_phase3_gr <- makeGRangesFromDataFrame(chr22_phased_highcov_phase3,
                         seqnames.field = "#CHROM",
                         start.field = "POS",
                         end.field = "POS",keep.extra.columns = T)

seqlevels(chr22_phased_highcov_phase3_gr) <- str_remove(seqlevels(chr22_phased_highcov_phase3_gr),"chr")
```

Some variants in 1000 genomes VCF are duplicated
```{r}
table(duplicated(chr22_phased_highcov_phase3_gr))
```

Exploring more of the duplicates

They seem to be due to the original VCF file reporting different variants in duplicate ranges, so for now I will skip these but might be a good idea to look more into this later on.

```{r}
chr22_phased_highcov_phase3_gr[duplicated(chr22_phased_highcov_phase3_gr) | duplicated(chr22_phased_highcov_phase3_gr,fromLast = TRUE)]
```

```{r}
chr22_phased_highcov_phase3_gr <- unique(chr22_phased_highcov_phase3_gr)

```


# Get overlaping variants from GWAS catalog in the vcf and get a list of variant ids

Not all match a variant in the vcf file, but most do
```{r}
distonear_gwascat_chr22 <- distanceToNearest(gwas_catalog_snp_gr_chr22_uq, chr22_phased_highcov_phase3_gr)
table(distonear_gwascat_chr22@elementMetadata$distance == 0)
```


Join overlapping ranges 
There is a 1-1 match between gwas snps and 1000 genomes vcf 
```{r}
gwas_100genomes_join <- plyranges::join_overlap_left(gwas_catalog_snp_gr_chr22_uq, chr22_phased_highcov_phase3_gr)

```

Print list to file to use as snpid in plink ld calculation
```{r eval = FALSE}
writeLines(gwas_100genomes_join$ID, "../data_output/plink/ld_gwas_1000genomesid_european_chr22.txt")
```

## Get list of unrelated European samples from 1000 genomes

List of original 2504 individuals
"http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel"

Split and write a file of unrelated samples for each population
```{r eval = FALSE}
unrelated_samples <- read_table("../data_raw/genomes_1000/integrated_call_samples_v3.20130502.ALL.panel")

unrelated_samples <- unrelated_samples[,c("sample","pop")]

unrelated_samples_list <- split(unrelated_samples[,c("sample","sample")], unrelated_samples$pop)

iwalk(unrelated_samples_list, 
     ~ write_tsv( x = .x,
                  file = paste0("../data_output/plink/sample_list_unrel_",.y,".txt"),col_names = FALSE))

```



# LD calculation with plink

Get LD estimates

Based on previous runs using the whole chromosome, the average ld block length is under 100kb, so we are using a 1MB window.

```{bash eval = FALSE}
vcf=../data_raw/genomes_1000/CCDG_14151_B01_GRM_WGS_2020-08-05_chr22.filtered.shapeit2-duohmm-phased.vcf.gz
out=../data_output/plink/chr22_european_gwas_snps_out_gbr_unrel
snp_list=../data_output/plink/ld_gwas_1000genomesid_european_chr22.txt
samples_list=../data_output/plink/sample_list_unrel_gbr.txt
plink --keep $samples_list --ld-snp-list $snp_list --vcf $vcf --ld-window-r2 0.2 --ld-window 5000000 --ld-window-kb 1000 --r2 --out $out


```

Read LD file
```{r}
chr22_all_gwas_snps_out_ld <- read_table("../data_output/plink/chr22_european_gwas_snps_out_gbr_unrel.ld")
```

# LD plots

## Average "LD decay length"
  
This plot shows the distributions of ranges between SNPs with r2 > 0.2 with GWAS SNPs

```{r}

chr22_all_gwas_snps_out_ld %>%
  group_by(SNP_A) %>%
  summarise(range1 = range(BP_B)[1],
            range2 = range(BP_B)[2]) %>%
  mutate(rangelen = range2 - range1) %>%
  ggplot(aes((rangelen+1)/1e3)) +
  geom_histogram(bins = 50, color = "white") +
  theme_light() +
  xlab("LD block length (kb)") +
  scale_x_log10()


``` 

This plot shows the distributions of ranges between SNPs with r2 > 0.7 with GWAS SNPs

```{r}

chr22_all_gwas_snps_out_ld %>%
  filter(R2 >= 0.7) %>%
  group_by(SNP_A) %>%
  summarise(range1 = range(BP_B)[1],
            range2 = range(BP_B)[2]) %>%
  mutate(rangelen = range2 - range1) %>%
  ggplot(aes((rangelen+1)/1e3)) +
  geom_histogram(bins = 50, color = "white") +
  theme_light() +
  xlab("LD block length (kb)") +
  scale_x_log10()


```

## Global LD decay 
  
```{r}
chr22_all_gwas_snps_out_ld %>%
ggplot(aes(abs(BP_A - BP_B)/1e3, R2)) +
  geom_bin2d( ) +
  scale_fill_gradient(low="lightgray",high="blue") +
  geom_smooth(color = "black") +
  theme_light() +
  theme(aspect.ratio = 1) +
  xlab("Distance (Kb)")
```
```{r}
chr22_all_gwas_snps_out_ld %>%
  filter(abs(BP_A -BP_B) < 100000) %>%
ggplot(aes(abs(BP_A - BP_B)/1e3, R2)) +
  geom_bin2d( ) +
  scale_fill_gradient(low="lightgray",high="blue") +
  geom_smooth(color = "black") +
  theme_light() +
  theme(aspect.ratio = 1) +
  xlab("Distance (Kb)")
```
## Example plots of surrounding variants in LD per SNP 

rs8141347

https://useast.ensembl.org/Homo_sapiens/Variation/LDPlot?db=core;pop1=373514;r=22:17660569-17661569;v=rs8141347;vdb=variation;vf=185116190

```{r}

bp_a_pos <- 17661069
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:17661069:C:T") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") +
  xlim(c( (bp_a_pos - 100000)/1e6 ,(bp_a_pos +  100000)/1e6 ))

```

rs138596453

https://useast.ensembl.org/Homo_sapiens/Variation/LDPlot?db=core;pop1=373514;r=22:17006086-17007086;v=rs138596453;vdb=variation;vf=187439569


```{r}

bp_a_pos <- 17006586
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:17006586:G:A") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") +
  xlim(c( (bp_a_pos - 100000)/1e6 ,(bp_a_pos +  100000)/1e6 ))

```

rs2072857
https://useast.ensembl.org/Homo_sapiens/Variation/LDPlot?db=core;pop1=373514;r=22:40420337-40421337;v=rs2072857;vdb=variation;vf=184514527

```{r}

bp_a_pos <- 40420837
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:40420837:G:A") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") +
  xlim(c( (bp_a_pos - 100000)/1e6 ,(bp_a_pos +  100000)/1e6 ))

```
# Filter by R2 > 0.7

Filtering for LD with R2 of at least 0.7
```{r}
chr22_all_gwas_snps_out_ld_filt <- chr22_all_gwas_snps_out_ld %>%
  filter(R2 >= 0.7)
```


# Overlap with haqers

Make GRanges

```{r}
chr22_all_gwas_snps_out_ld_filt_gr <- makeGRangesFromDataFrame(df = chr22_all_gwas_snps_out_ld_filt,
                         seqnames.field = "CHR_B",
                         start.field = "BP_B",
                         end.field = "BP_B",
                         keep.extra.columns = TRUE)
```


```{r}
chr22_all_gwas_snps_out_ld_filt_gr_haqers <- plyranges::join_overlap_inner(haqers,chr22_all_gwas_snps_out_ld_filt_gr)
```



Number of haqers that overlap at least one snp
```{r}
unique(chr22_all_gwas_snps_out_ld_filt_gr_haqers$name)
```

Haqers width
```{r}
width(unique(chr22_all_gwas_snps_out_ld_filt_gr_haqers))
```

Number of SNPs per haqer
```{r}

chr22_all_gwas_snps_out_ld_filt_gr_haqers_df <- as.data.frame(elementMetadata(chr22_all_gwas_snps_out_ld_filt_gr_haqers))

lapply(split(chr22_all_gwas_snps_out_ld_filt_gr_haqers_df,chr22_all_gwas_snps_out_ld_filt_gr_haqers_df$name), function(x){
  length(unique(x$SNP_B))
})
```
```{r}
lapply(split(chr22_all_gwas_snps_out_ld_filt_gr_haqers_df,chr22_all_gwas_snps_out_ld_filt_gr_haqers_df$name), function(x){
  length(unique(x$SNP_A))
})
```

```{r}
table(chr22_all_gwas_snps_out_ld_filt_gr_haqers$SNP_A,chr22_all_gwas_snps_out_ld_filt_gr_haqers$SNP_B)

unique(chr22_all_gwas_snps_out_ld_filt_gr_haqers$SNP_A)
unique(chr22_all_gwas_snps_out_ld_filt_gr_haqers$SNP_B)
```

# Phenotypes of associated haqers

Get id info 
```{r}
haqer_snps_df <- left_join(as.data.frame(chr22_all_gwas_snps_out_ld_filt_gr_haqers), as.data.frame(gwas_100genomes_join), by = c("SNP_A" = "ID"))

haqer_snps <- unique(haqer_snps_df[,c("SNPS","name")])
```


```{r}
lapply(split(haqer_snps$SNPS,haqer_snps$name), function(x, df){
  df <- df[df$SNPS %in% x,]
  table(df$`DISEASE/TRAIT`)

}, df = gwas_catalog_snp)
```


```{r}
gwas_catalog_snp_haqers <- gwas_catalog_snp[gwas_catalog_snp$SNPS %in% haqer_snps$SNPS,]
```

```{r}
table(gwas_catalog_snp_haqers$`DISEASE/TRAIT`)
```

How many GWAS SNPs have we gained by using LD snps

```{r}
chr22_all_gwas_snps_out_ld_filt_gr_haqers <- plyranges::join_overlap_inner(haqers,gwas_catalog_snp_gr_chr22)
```

Actually all
```{r}
chr22_all_gwas_snps_out_ld_filt_gr_haqers
```

# Plot R2 variants for HAQER overlaps
```{r}
chr22_all_gwas_snps_out_ld_filt_gr_haqers

chr22_all_gwas_snps_out_ld
```

```{r}

bp_a_pos <- 26037088
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:26037088:C:T") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_vline(xintercept = 26348220 / 1e6, col = "green") +
  geom_vline(xintercept = 26349068 / 1e6, col = "lightblue") +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") 
  

```
```{r}

bp_a_pos <- 40876139
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:40876139:T:C") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_vline(xintercept = 40974367 / 1e6, col = "gray") +
  geom_vline(xintercept = 40975378 / 1e6, col = "lightblue") +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") 
  

```
```{r}

bp_a_pos <- 49612430
chr22_all_gwas_snps_out_ld %>%
  filter(SNP_A == "22:49612430:C:T") %>%
ggplot(aes(BP_B / 1e6, R2)) +
  geom_vline(xintercept = 49595392 / 1e6, col = "gray") +
  geom_vline(xintercept = 49596243 / 1e6, col = "lightblue") +
  geom_point() +
  geom_vline(xintercept = bp_a_pos / 1e6)  +
  geom_hline(yintercept = 0.7,lty = "dashed", color = "red") +
  theme_light() +
  xlab("Position (Mb)") 
  

```
