---
title: "gwas_catalog_exploration"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)

```

# Read gwas catalog data

```{r}
gwas_catalog_ancestry <- read_tsv("../data_raw/gwas_catalog/gwas_catalog-ancestry_r2021-12-21.tsv")
```

# Exploration of ancestries
```{r}
gwas_catalog_ancestry <- gwas_catalog_ancestry %>%
  mutate(broad_ancestral_category = `BROAD ANCESTRAL CATEGORY`) 

```

How many studies per ancestry are there for the top 10 most abundant single ancestry studies?
```{r}
gwas_catalog_ancestry %>%
  count(broad_ancestral_category,name = "ancestry_count") %>%
  top_n(n = 10) %>%
  mutate(broad_ancestral_category = fct_reorder(broad_ancestral_category,ancestry_count)) %>%
  ggplot(aes(broad_ancestral_category,ancestry_count, fill = broad_ancestral_category)) +
  geom_col() +
  coord_flip() +
  ylab(NULL) +
  theme_minimal() +
  theme(aspect.ratio = 1,legend.position = "none") +
  scale_fill_brewer(type = "qual",palette = 3)
```

How many studies per ancestry are there for the top 10 most abundant single ancestry studies, taking out European?

```{r}
gwas_catalog_ancestry %>%
  count(broad_ancestral_category,name = "ancestry_count") %>%
  top_n(n = 10) %>%
  mutate(broad_ancestral_category = fct_reorder(broad_ancestral_category,ancestry_count)) %>%
  filter(broad_ancestral_category != "European") %>%
  ggplot(aes(broad_ancestral_category,ancestry_count, fill = broad_ancestral_category)) +
  geom_col() +
  coord_flip() +
  ylab(NULL) +
  theme_minimal() +
  theme(aspect.ratio = 1,legend.position = "none") +
  scale_fill_brewer(type = "qual",palette = 3)
```

# Explore SNP data

```{r}
gwas_catalog_snp <- read_tsv("../data_raw/gwas_catalog/gwas_catalog_v1.0.2-associations_e105_r2021-12-21.tsv")
```

```{r}
gwas_catalog_snp %>%
  count(MAPPED_TRAIT,name = "trait_count") %>%
  top_n(n = 20) %>%
  mutate(MAPPED_TRAIT = fct_reorder(MAPPED_TRAIT,trait_count)) %>%
  ggplot(aes(MAPPED_TRAIT,trait_count, fill = MAPPED_TRAIT)) +
  geom_col() +
  coord_flip() +
  ylab(NULL) +
  theme_minimal() +
  theme(aspect.ratio = 1,legend.position = "none") +
  scale_fill_viridis_d(begin= 0.3,end = 0.6) +
  ylab("Number of SNP associations") +
  xlab(NULL)
```
```{r}
haqers_olap_pheno_df <- readRDS("../data_output/rds/haqers_olap_pheno_df.Rds")

unique(haqers_olap_pheno_df[,c("name","MAPPED_TRAIT")]) %>%
  count(MAPPED_TRAIT,name = "trait_count") %>%
  #top_n(n = 20) %>%
  mutate(MAPPED_TRAIT = fct_reorder(MAPPED_TRAIT,trait_count)) %>%
  ggplot(aes(MAPPED_TRAIT,trait_count, fill = MAPPED_TRAIT)) +
  geom_col() +
  coord_flip() +
  ylab(NULL) +
  theme_minimal() +
  theme(aspect.ratio = 1,legend.position = "none") +
  scale_fill_viridis_d(begin= 0.3,end = 0.5,option = "B") +
  ylab("Number of HAQER associations") +
  xlab(NULL)
```


```{r}
gwas_catalog_snp %>%
  mutate(DISEASE_TRAIT = `DISEASE/TRAIT`) %>%
  count(DISEASE_TRAIT,name = "trait_count") %>%
  top_n(n = 50) %>%
  mutate(DISEASE_TRAIT = fct_reorder(DISEASE_TRAIT,trait_count)) %>%
  ggplot(aes(DISEASE_TRAIT,trait_count, fill = DISEASE_TRAIT)) +
  geom_col() +
  coord_flip() +
  ylab(NULL) +
  theme_minimal() +
  theme(aspect.ratio = 1,legend.position = "none") +
  scale_fill_viridis_d()
```
How many traits have been mapped to EFO
```{r}
summary(is.na(gwas_catalog_snp$`DISEASE/TRAIT`))
```
Only 836 are missing

There are multiple ancestries per study!
```{r}
rowSums(table(gwas_catalog_ancestry$`STUDY ACCESSION`,gwas_catalog_ancestry$broad_ancestral_category)) %>%
  enframe(name = "study_accession",value = "count") %>%
  ggplot(aes(count)) +
  geom_histogram()
  
```

For now focus on single ancestry studies
```{r}


gwas_catalog_ancestry_nodups <- gwas_catalog_ancestry[!(duplicated(gwas_catalog_ancestry$`STUDY ACCESSION`) | duplicated(gwas_catalog_ancestry$`STUDY ACCESSION`, fromLast = TRUE)),]

```

Join SNPs and Ancestries
```{r}
gwas_catalog_snp <- gwas_catalog_snp %>%
  select("CHR_ID","CHR_POS","SNP_ID_CURRENT","DISEASE/TRAIT","STUDY ACCESSION")

gwas_catalog_ancestry_nodups <- gwas_catalog_ancestry_nodups %>%
  select("broad_ancestral_category","ADDITONAL ANCESTRY DESCRIPTION","STUDY ACCESSION")

gwas_snp_ancestry<- left_join(gwas_catalog_snp, gwas_catalog_ancestry_nodups,by = "STUDY ACCESSION")

```

# Quick exploration with HAQERs

Read haqers
```{r}
library(rtracklayer)
haqers <- import.bed("../data_raw/bed/HAQER.bed")
```



Get SNPs as GRanges
```{r}
gwas_snp_ancestry$CHR_POS <- as.numeric(gwas_snp_ancestry$CHR_POS)

gwas_snp_ancestry <- gwas_snp_ancestry[!is.na(gwas_snp_ancestry$CHR_POS),]

gwas_snp_ancestry_granges <- GRanges(seqnames = gwas_snp_ancestry$CHR_ID,
        ranges = IRanges(start = gwas_snp_ancestry$CHR_POS,width = 1))

elementMetadata(gwas_snp_ancestry_granges) <- DataFrame(gwas_snp_ancestry)
```


```{r}
library(plyranges)
```

```{r}
seqlevels(haqers) <- str_remove(seqlevels(haqers),"chr")


haqers$name <- paste0("haqer",1:length(haqers))
haqers_snp_ancestry <- plyranges::join_overlap_inner_within(gwas_snp_ancestry_granges, haqers)
```

```{r}
haqers_snp_ancestry %>%
  as.data.frame() %>%
  ggplot(aes(DISEASE.TRAIT)) +
  geom_bar() +
  coord_flip() +
  theme_minimal()
```
```{r}
haqers_snp_ancestry %>%
  as.data.frame() %>%
  ggplot(aes(broad_ancestral_category)) +
  geom_bar() +
  coord_flip() +
  theme_minimal()

```

```{r}

A <- GRanges(seqnames = "chr1",
        ranges = IRanges(10,width = 7))
B <- GRanges(seqnames = "chr1",
        ranges= IRanges(5,15))
join_overlap_inner_within(A, B)

```


```{r}
countOverlaps(haqers,gwas_snp_ancestry_granges) %>%
  enframe() %>%
  ggplot(aes(value)) +
  geom_histogram(bins = 5)
```

```{r}
haqers_dist_to_nearest_snp <- distanceToNearest(haqers,gwas_snp_ancestry_granges)

haqers_dist_to_nearest_snp@elementMetadata$distance %>%
  enframe() %>%
  ggplot(aes(value)) +
  geom_histogram(bins = 100)


summary(haqers_dist_to_nearest_snp@elementMetadata$distance < 100)
```

```{r}
length(unique(haqers_snp_ancestry$name))
```

```{r}
median(colSums(table(haqers_snp_ancestry$SNP_ID_CURRENT,haqers_snp_ancestry$name)))
```
```{r}
haqers_snp_ancestry$DISEASE.TRAIT
```


```{r}
haqers_snp_ancestry %>%
  as.data.frame() %>%
  pull(DISEASE.TRAIT) %>%
  table() %>%
  enframe() %>%
  mutate(name = fct_reorder(name, value)) %>%
  top_n(20, value) %>%
  ggplot(aes(value, name)) +
  geom_col() +
  theme_light()
```

```{r}
haqers_snp_ancestry %>%
  as.data.frame() %>%
  pull(DISEASE.TRAIT) %>%
  table() %>%
  enframe() %>%
  mutate(name = fct_reorder(name, value)) %>%
  ggplot(aes(value, name)) +
  geom_col() +
  theme_light()
```

# Studies table

```{r}
gwas_studies <- read_tsv("../data_raw/gwas_catalog/gwas-catalog-v1.0.3-studies-r2022-01-12.tsv")
```

```{r}
all(is.na(gwas_studies$`SUMMARY STATS LOCATION`))
```

# Summary statistics info

```{r}
list_gwas_summary_statistics <- read_csv("../data_raw/gwas_catalog/list_gwas_summary_statistics_28_Jan_2022.csv")
```


```{r}
sum(gwas_catalog_snp$`STUDY ACCESSION` %in% list_gwas_summary_statistics$`Study accession`) / length(gwas_catalog_snp$`STUDY ACCESSION`)
```
Example summary statistics page

```{r}
GCST001969_EFO_0004326 <- read_tsv("../data_raw/gwas_catalog/23583979-GCST001969-EFO_0004326.h.tsv.gz")

```

```{r}
GCST001969_EFO_0004326 %>%
  slice_sample(n = 1e4) %>%
  ggplot(aes(base_pair_location,- 10 * log10(p_value_gcadj))) +
    geom_point() +
    facet_grid(cols = vars(chromosome))
```

```{r}
GCST001969_EFO_0004326_gwas_snps <- GCST001969_EFO_0004326[GCST001969_EFO_0004326$hm_rsid %in% gwas_catalog_snp[gwas_catalog_snp$`STUDY ACCESSION` == "GCST001969" & gwas_catalog_snp$CHR_ID == "1",]$SNPS,]

GCST001969_EFO_0004326_ld_snps<- out_ld_df[out_ld_df$CHR_A %in% GCST001969_EFO_0004326_gwas_snps$chromosome &
out_ld_df$BP_A %in% GCST001969_EFO_0004326_gwas_snps$base_pair_location,]

GCST001969_EFO_0004326_ld_snps_range <- range(GCST001969_EFO_0004326_ld_snps$BP_B)


GCST001969_EFO_0004326 %>%
  filter(chromosome == "1") %>%
  ggplot(aes(base_pair_location,- 10 * log10(p_value_gcadj))) +
    geom_point() +
  geom_point(data = GCST001969_EFO_0004326_gwas_snps,color = "red") +
  geom_vline(xintercept = GCST001969_EFO_0004326_ld_snps_range[1], color = "darkblue") +
  geom_vline(xintercept = GCST001969_EFO_0004326_ld_snps_range[2], color = "darkblue") +
    geom_vline(xintercept = haqers_start_chr1[1], color = "lightblue") +
  geom_vline(xintercept = haqers_start_chr1[2], color = "lightblue") +
  xlim(c(205e6,210e6)) 
  


sum(gwas_catalog_snp$`STUDY ACCESSION` == "GCST001969")


```

```{r}
plyranges::join_overlap_left(haqers, GRanges(seqnames = "1",
        ranges = IRanges(start = GCST001969_EFO_0004326_ld_snps_range[1],end = GCST001969_EFO_0004326_ld_snps_range[2])))

haqers_start_chr1 <- start(haqers[seqnames(haqers)==1])
haqers_start_chr1 <- haqers_start_chr1[dplyr::between(haqers_start_chr1,left = 205e6,right = 210e6)]
```



