---
title: "phenotype_enrichments_downstream_steps"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(rtracklayer)
library(plyranges)

options(scipen = 9999)

```

# Functions
```{r}
# Count haqers per phenotye 
get_haqers_per_pheno <- function(.x,.y) {
  haqers_per_pheno <- as.data.frame(table(unique(.x[,c("name","MAPPED_TRAIT")])$MAPPED_TRAIT))
  haqers_per_pheno$n_rand <- .y
  return(haqers_per_pheno)
}
# Calc p value
get_pheno_pval <- function(.x, .y, random_counts_df){
  random_counts <- random_counts_df[random_counts_df$mapped_trait == .x,"counts"]
  nrand <- length(unique(random_counts_df$id_rand))
  pval <- (sum(random_counts >= .y ) +1) / (nrand+1)
  return(pval)
}


```

# Read overlaps data

```{r}
gwas_ld_snp_random_pheno_df_list <- readRDS("../data_output/rds/gwas_ld_snp_random_pheno_df_list.Rds")
haqers_olap_pheno_df <- readRDS("../data_output/rds/haqers_olap_pheno_df.Rds")
```

# Count overlaps

Get number of haqers per phenotype for each random haqer
```{r}

names(gwas_ld_snp_random_pheno_df_list) <- paste0("rand_",1:length(gwas_ld_snp_random_pheno_df_list))

haqers_per_pheno_rand_list <- imap(gwas_ld_snp_random_pheno_df_list,get_haqers_per_pheno)
```

Bind all 
```{r}
haqers_per_pheno_rand_df <- do.call(rbind,haqers_per_pheno_rand_list)
colnames(haqers_per_pheno_rand_df) <- c("mapped_trait","counts","id_rand")

```

Get same info for actual haqers

```{r}
haqers_per_pheno_counts <- get_haqers_per_pheno(haqers_olap_pheno_df,"haqers")

colnames(haqers_per_pheno_counts) <- c("mapped_trait","counts","id_haqer")
haqers_per_pheno_counts <- select(haqers_per_pheno_counts, -id_haqer)

```

### Get counts by gwas snp instead of haqer
```{r}
get_gwas_per_pheno <- function(.x,.y) {
  haqers_per_pheno <- as.data.frame(table(unique(.x[,c("SNPS","MAPPED_TRAIT")])$MAPPED_TRAIT))
  haqers_per_pheno$n_rand <- .y
  return(haqers_per_pheno)
}
gwas_per_pheno_counts <- get_gwas_per_pheno(haqers_olap_pheno_df,"haqers")
colnames(gwas_per_pheno_counts) <- c("mapped_trait","counts","id_haqer")
gwas_per_pheno_counts <- select(gwas_per_pheno_counts, -id_haqer)

```
Get number of haqers per phenotype for each random haqer
```{r}

gwas_per_pheno_rand_list <- imap(gwas_ld_snp_random_pheno_df_list,get_gwas_per_pheno)
```

Bind all 
```{r}
gwas_per_pheno_rand_df <- do.call(rbind,gwas_per_pheno_rand_list)
colnames(gwas_per_pheno_rand_df) <- c("mapped_trait","counts","id_rand")

```


# Get "p value" per phenotype

```{r}
# Convert factors to characters for comparison
haqers_per_pheno_rand_df$mapped_trait <- as.character(haqers_per_pheno_rand_df$mapped_trait)
haqers_per_pheno_counts$mapped_trait <- as.character(haqers_per_pheno_counts$mapped_trait)



haqers_per_pheno_counts$pval <- map2_dbl(as.character(haqers_per_pheno_counts$mapped_trait),
         haqers_per_pheno_counts$counts,
         get_pheno_pval,
         random_counts_df = haqers_per_pheno_rand_df)

haqers_per_pheno_counts$fdr <- p.adjust(haqers_per_pheno_counts$pval,method = "fdr")
```
```{r}
write_tsv(haqers_per_pheno_counts, "../data_output/tsv/mapped_trait_haqers_pval_1000.tsv")

```
```{r}
# Convert factors to characters for comparison
gwas_per_pheno_rand_df$mapped_trait <- as.character(gwas_per_pheno_rand_df$mapped_trait)
gwas_per_pheno_counts$mapped_trait <- as.character(gwas_per_pheno_counts$mapped_trait)



gwas_per_pheno_counts$pval <- map2_dbl(as.character(gwas_per_pheno_counts$mapped_trait),
         gwas_per_pheno_counts$counts,
         get_pheno_pval,
         random_counts_df = gwas_per_pheno_rand_df)

gwas_per_pheno_counts$fdr <- p.adjust(gwas_per_pheno_counts$pval,method = "fdr")
```



# Exploration 


Distribution of p values

Counting with haqers
```{r}
haqers_per_pheno_counts %>%
  ggplot(aes(pval)) +
  geom_histogram()
```

Counting with snps
```{r}
gwas_per_pheno_counts %>%
  ggplot(aes(pval)) +
  geom_histogram()

```


Phenotypes
```{r}
haqers_per_pheno_counts_fdr5 <- haqers_per_pheno_counts[haqers_per_pheno_counts$fdr< 0.05,] %>%
  arrange(fdr) 
```

```{r}
gwas_per_pheno_counts_fdr5 <- gwas_per_pheno_counts[gwas_per_pheno_counts$fdr< 0.05,] %>%
  arrange(fdr) 

sum(gwas_per_pheno_counts$pval < 0.05)

```

Independent filtering improves power?
```{r}
gwas_per_pheno_counts_sub <- gwas_per_pheno_counts[gwas_per_pheno_counts$counts > 2,]
gwas_per_pheno_counts_sub$fdr <- p.adjust(gwas_per_pheno_counts_sub$pval,method = "fdr")
sum(gwas_per_pheno_counts_sub$fdr < 0.1)

gwas_per_pheno_counts_sub[gwas_per_pheno_counts_sub$fdr < 0.1,]
```


Phenotypes from other subset filtered by width

```{r}
haqers_per_pheno_counts_filtwidth <- read_tsv("../data_output/tsv/mapped_trait_haqers_pval_1000_filtwidth.tsv")
haqers_per_pheno_counts_filtwidth$fdr <- p.adjust(haqers_per_pheno_counts_filtwidth$pval,method = "fdr")


haqers_per_pheno_counts_filtwidth_fdr10 <- haqers_per_pheno_counts_filtwidth[haqers_per_pheno_counts_filtwidth$fdr < 0.1,] %>%
  arrange(fdr) 

```

```{r}
traits_filtwidth <- haqers_per_pheno_counts_fdr5$mapped_trait[haqers_per_pheno_counts_fdr5$mapped_trait %in% 
haqers_per_pheno_counts_filtwidth_fdr10$mapped_trait]


traits_remaining <- haqers_per_pheno_counts_fdr5$mapped_trait[!haqers_per_pheno_counts_fdr5$mapped_trait %in% 
haqers_per_pheno_counts_filtwidth_fdr10$mapped_trait]


traits_filtwidth[traits_filtwidth %in% gwas_per_pheno_counts$mapped_trait[gwas_per_pheno_counts$fdr < 0.2]]

```

```{r}
haqers_olap_pheno_df_schz <-haqers_olap_pheno_df[haqers_olap_pheno_df$MAPPED_TRAIT == "unipolar depression, schizophrenia, sex interaction measurement, bipolar disorder",]


unique(haqers_olap_pheno_df_schz$SNPS)
```

```{r}
str_split(unique(haqers_olap_pheno_df_schz$SNP_B)[-1],":",)

haqers_olap_pheno_df_schz[-c(1:4),]

haqers_olap_pheno_df_schz %>%
  filter(!across(everything(),is.na)) %>%
  mutate(CHR_B = str_split(SNP_B,":") %>% map_chr(1),
         BP_B = str_split(SNP_B,":") %>% map_chr(2)) %>%
  select( CHR_B, BP_B) %>%
  mutate(CHR_B = paste0("chr",CHR_B)) %>%
  unique() %>%
  rownames_to_column() %>%
  makeGRangesFromDataFrame(seqnames.field = "CHR_B",start.field =  "BP_B",end.field = "BP_B") %>%
  export.bed(con = "../data_clean/ld_snps_schizophrenia.bed")
```


# Plots


```{r}
trait <- "unipolar depression, schizophrenia, sex interaction measurement, bipolar disorder"


plot_pheno_enrichment <- function(trait, random_traits_df, haqer_traits_df) {
  random_traits_df %>%
  filter(mapped_trait == trait) %>%
  ggplot(aes(counts)) +
  geom_histogram() +
  geom_vline(xintercept = haqer_traits_df %>%
               filter(mapped_trait == trait) %>%
               pull(counts),
             lty = "dashed",
             color = "red") +
    theme_light() +
    theme(aspect.ratio = 0.5) +
    ggtitle(trait) 

}

plot_pheno_enrichment(trait,random_traits_df = haqers_per_pheno_rand_df,haqers_per_pheno_counts)
haqers_per_pheno_counts$mapped_trait
```

```{r}
map(traits_filtwidth, plot_pheno_enrichment,random_traits_df = haqers_per_pheno_rand_df,haqers_per_pheno_counts )
```

```{r}
map(traits_remaining, plot_pheno_enrichment,random_traits_df = haqers_per_pheno_rand_df,haqers_per_pheno_counts )
```


## More info on traits

- What are the original traits
```{r}

traits_filtwidth_original <- haqers_olap_pheno_df[haqers_olap_pheno_df$MAPPED_TRAIT %in% traits_filtwidth,c("MAPPED_TRAIT","DISEASE/TRAIT")]

lapply(split(traits_filtwidth_original$`DISEASE/TRAIT`,traits_filtwidth_original$MAPPED_TRAIT),unique)
```

```{r}
traits_remaining_original <- haqers_olap_pheno_df[haqers_olap_pheno_df$MAPPED_TRAIT %in% traits_remaining,c("MAPPED_TRAIT","DISEASE/TRAIT")]

lapply(split(traits_remaining_original$`DISEASE/TRAIT`,traits_remaining_original$MAPPED_TRAIT),unique)

```


- Number of chromosomes
```{r}
traits_filtwidth_chrs <- haqers_olap_pheno_df[haqers_olap_pheno_df$MAPPED_TRAIT %in% traits_filtwidth,c("MAPPED_TRAIT","seqnames", "SNPS")]
traits_remaining_chrs <- haqers_olap_pheno_df[haqers_olap_pheno_df$MAPPED_TRAIT %in% traits_remaining,c("MAPPED_TRAIT","seqnames", "SNPS")]


lapply(split(traits_filtwidth_chrs$`seqnames`,traits_filtwidth_chrs$MAPPED_TRAIT),unique)
traits_remaining_chrs_uq <- lapply(split(traits_remaining_chrs$`seqnames`,traits_remaining_chrs$MAPPED_TRAIT),unique)

```
- Number of GWAS SNP

```{r}
lapply(split(traits_filtwidth_chrs$`SNPS`,traits_filtwidth_chrs$MAPPED_TRAIT),function(x) length(unique(x)))
traits_remaining_nsnps <- lapply(split(traits_remaining_chrs$`SNPS`,traits_remaining_chrs$MAPPED_TRAIT),function(x) length(unique(x)))

```

```{r}
for (name in names(traits_remaining_nsnps)){
  cat(paste(name,"\n","Chrs:\n", paste(traits_remaining_chrs_uq[[name]],collapse = " "),"\nNumber of SNPs:\n",traits_remaining_nsnps[[name]],"\n\n", collapse = " "))
}
name
traits_remaining_chrs_uq[name]
```


```{r}
traits_filtwidth_chrs %>%
  filter(str_detect(MAPPED_TRAIT,"schi")) %>%
  pull(SNPS) %>%
  unique()
```

```{r}
traits_remaining_chrs %>%
  filter(str_detect(MAPPED_TRAIT,"remission")) %>%
  pull(SNPS) %>%
  unique()


```

# SNPs per 
```{r}
gwas_catalog_snp <- read_tsv("../data_raw/gwas_catalog/gwas_catalog_v1.0.2-associations_e105_r2021-12-21.tsv")


table(gwas_catalog_snp[gwas_catalog_snp$MAPPED_TRAIT == "unipolar depression", na.rm = T,]$CHR_ID)
```


